rfd/0138-postgres-backend.md

```diff
@@ -0,0 +1,132 @@
+
+The `event_id` column is a UUID chosen at random at insert time, unrelated to the event data, only used to disambiguate events for the purpose of pagination; `creation_time`, likewise, is only used for deleting events after some retention period (defaulting to one year).
+
+With minor modifications, this schema can support range based partitioning by `event_time`, creating partitions by hand or using `pg_partman`. The initial implementation will not use partitioning, but setting up the table layout for it will allow for a much easier migration if and when we end up supporting it.
+
+### Requirements and recommendations
+
+Any supported version of Postgres (11 to 15) should work, with `wal2json` 2.1 or later, although for better results, Postgres 13 or higher is recommended (as logical decoding had some internal improvements).
+
+Because the db user that Teleport connects as needs `REPLICATION` permissions, which effectively grants read access to the entire database cluster, the backend should not be used in a database cluster that also contains data that the Teleport auth server should not have access to (it's fine to use a single cluster for the backend and the audit log storage, for instance).
+
+Because of how temporary replication slots (and prepared statements) are tied to individual connections, the use of external connection poolers (such as `pgbouncer`) is not supported.
+
+For self-managed Postgres setups we recommend the use of [certificate authentication](https://www.postgresql.org/docs/current/auth-cert.html); in Azure we support and recommend the use of Azure AD authentication over the use of hardcoded passwords.
+
+Teleport will attempt to create the databases specified in the configuration for itself, and set them up with the correct schema (as well as apply any future migrations, if needed). It's possible to create the databases manually and not give Teleport permissions to create databases, and technically it's even possible to manually set up the database schemas (to avoid granting Teleport the ability to create tables, for instance), but we recommend giving Teleport its own database for the backend - and since Teleport will need read-write access to all the data in it anyway, we also recommend letting Teleport manage the schema automatically.
+
+It's possible to only grant `INSERT` and `SELECT` privileges to the Teleport db user on the table that stores audit events, but not UPDATE or DELETE privileges; such a setup couldn't be automatically managed by Teleport, however, and thus would require detailed documentation on how to set up the schema, and some care.
```

Comment:
Have you considered taking a hybrid approach, where the Teleport binary, run with a special setup command, performs everything for the ops person, and then, once ran in the production setup, can operate with lower privileges? (If it's a moot point because of the replication anyway, feel free to ignore this question.)

Reply:
Replication permissions only allow reading from the database cluster, and Teleport needs full read access over both the backend and the whole audit log anyway. I think I'd rather provide some shell script (or a list of commands to copy and paste?) that uses  `psql`, which can be more easily run on any computer - including, conveniently, the Azure cloud shell prompt.

Reply:
That being said, if the instructions end up being too onerous it's certainly a thing we could consider!